flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java

```diff
@@ -0,0 +1,68 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.pulsar.source;
+
+import org.apache.flink.connector.pulsar.testutils.PulsarContainerContextFactory;
+import org.apache.flink.connector.pulsar.testutils.PulsarContainerEnvironment;
+import org.apache.flink.connector.pulsar.testutils.cases.MultipleTopicConsumingContext;
+import org.apache.flink.connector.pulsar.testutils.cases.SingleTopicConsumingContext;
+import org.apache.flink.connectors.test.common.environment.MiniClusterTestEnvironment;
+import org.apache.flink.connectors.test.common.junit.annotations.ExternalContextFactory;
+import org.apache.flink.connectors.test.common.junit.annotations.ExternalSystem;
+import org.apache.flink.connectors.test.common.junit.annotations.TestEnv;
+import org.apache.flink.connectors.test.common.testsuites.SourceTestSuiteBase;
+
+import org.junit.jupiter.api.Nested;
+
+/** Unite test class for {@link PulsarSource}. */
+@SuppressWarnings("unused")
+class PulsarSourceITCase {
+
+    @Nested
+    class SingleTopicConsuming extends SourceTestSuiteBase<String> {
+
+        // Defines test environment on Flink MiniCluster
+        @TestEnv MiniClusterTestEnvironment flink = new MiniClusterTestEnvironment();
+
+        // Defines pulsar running environment
+        @ExternalSystem PulsarContainerEnvironment pulsar = new PulsarContainerEnvironment();
+
+        // Defines a external context Factories,
+        // so test cases will be invoked using this external contexts.
+        @ExternalContextFactory
+        PulsarContainerContextFactory<String, SingleTopicConsumingContext> factory =
+                new PulsarContainerContextFactory<>(pulsar, SingleTopicConsumingContext.class);
+    }
+
+    @Nested
+    class MultipleTopicConsuming extends SourceTestSuiteBase<String> {
+
+        // Defines test environment on Flink MiniCluster
+        @TestEnv MiniClusterTestEnvironment flink = new MiniClusterTestEnvironment();
+
+        // Defines pulsar running environment
+        @ExternalSystem PulsarContainerEnvironment pulsar = new PulsarContainerEnvironment();
+
+        // Defines a external context Factories,
+        // so test cases will be invoked using this external contexts.
+        @ExternalContextFactory
+        PulsarContainerContextFactory<String, MultipleTopicConsumingContext> factory =
+                new PulsarContainerContextFactory<>(pulsar, MultipleTopicConsumingContext.class);
+    }
+}
```

Comment:
Could be
```suggestion
class PulsarSourceITCase extends SourceTestSuiteBase<String> {

    // Defines test environment on Flink MiniCluster
    @TestEnv MiniClusterTestEnvironment flink = new MiniClusterTestEnvironment();

    // Defines pulsar running environment
    @ExternalSystem PulsarContainerEnvironment pulsar = new PulsarContainerEnvironment();

    // Defines a external context Factories,
    // so test cases will be invoked using this external contexts.
    @ExternalContextFactory
    PulsarContainerContextFactory<String, MultipleTopicConsumingContext> multipleTopic =
            new PulsarContainerContextFactory<>(pulsar, MultipleTopicConsumingContext.class);

    // Defines a external context Factories,
    // so test cases will be invoked using this external contexts.
    @ExternalContextFactory
    PulsarContainerContextFactory<String, SingleTopicConsumingContext> singleTopic =
            new PulsarContainerContextFactory<>(pulsar, SingleTopicConsumingContext.class);
}
```

But I also like your current solution.